wcfsetup/install/files/lib/system/session/SessionHandler.class.php

```diff
@@ -405,46 +370,114 @@ public function getUser() {
 	}
 	
 	/**
-	 * Tries to read existing session identified by the given session id.
-	 * 
-	 * @param	string		$sessionID
+	 * Tries to read existing session identified by the given session id. Returns whether
+	 * a session could be found.
 	 */
-	protected function getExistingSession($sessionID) {
-		$this->session = new $this->sessionClassName($sessionID);
-		if (!$this->session->sessionID) {
-			$this->session = null;
-			return;
+	protected function getExistingSession(string $sessionID): bool {
+		$sql = "SELECT	*
+			FROM	wcf".WCF_N."_".($this->isACP ? 'acp' : 'user')."_session
+			WHERE	sessionID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$sessionID
+		]);
+		$row = $statement->fetchSingleRow();
+		
+		if (!$row) {
+			return false;
+		}
+		
+		// Check whether the session technically already expired.
+		$lifetime = ($this->isACP ? self::ACP_SESSION_LIFETIME :
+			($row['userID'] ? self::USER_SESSION_LIFETIME :
+			self::GUEST_SESSION_LIFETIME));
+		if ($row['lastActivityTime'] < (TIME_NOW - $lifetime)) {
+			return false;
+		}
+		
+		$this->sessionID = $sessionID;
+		$this->user = new User($row['userID']);
+		$this->variables = unserialize($row['sessionVariables']);
+		
+		$sql = "UPDATE	wcf".WCF_N."_".($this->isACP ? 'acp' : 'user')."_session
+			SET	ipAddress = ?,
+				userAgent = ?,
+				lastActivityTime = ?
+			WHERE	sessionID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			UserUtil::getIpAddress(),
+			UserUtil::getUserAgent(),
+			TIME_NOW,
+			$this->sessionID,
+		]);
+		
+		// Refresh cookie.
+		if ($this->user->userID) {
+			HeaderUtil::setCookie(($this->isACP ? 'acp' : 'user')."_session", $this->sessionID, TIME_NOW + 86400 * 14);
 		}
 		
-		$this->user = new User($this->session->userID);
+		// Fetch legacy session.
+		if (!$this->isACP) {
+			$sql = "SELECT	*
+				FROM	wcf".WCF_N."_session
+				WHERE	(userID IS NULL AND sessionID = ?)
+				OR	(userID IS NULL AND spiderID = ?)
+				OR	(userID IS NOT NULL AND userID = ?)";
```

Comment:
This query is quite awkward.

Reply:
Elaborate, please? Is it the formatting? Is it the `WHERE` condition?

Reply:
The query is quite generic because it tries all kinds of conditions, even though we know much more ahead of time. If `$row['userID']` does not contain a user id, then this is either a search engine or a regular guest. If it contains an user id, we don't care for the first two conditions at all.

Reply:
So I basically add an `if ($row['userID'])` to the PHP code and send a different query in the *then* and *else* part, yes?

Reply:
The logic goes like this:

1. If `$row['userID']` then only check for sessions with an exact match for `userID`.
2. If `$this->getSpiderID(…)` returns an id, then only check by `spiderID`.
3. Else check for the session id.

It might make sense to use a condition builder here for ease of logic.

Reply:
I've split this into user / guest only. The query optimizer will handle the `spiderID = NULL` case just fine.